File: test/timer_calib.c
/* Copyright 2012 The ChromiumOS Authors
 * Use of this source code is governed by a BSD-style license that can be
 * found in the LICENSE file.
 *
 * Tasks for scheduling test.
 */

#include "common.h"
#include "console.h"
#include "task.h"
#include "timer.h"
#include "util.h"

uint32_t difftime(timestamp_t t0, timestamp_t t1)
{
	return (uint32_t)(t1.val - t0.val);
}

int timer_calib_task(void *data)
{
	timestamp_t t0, t1;
	unsigned int d;

	while (1) {
		task_wait_event(-1);

		ccprintf("\n=== Timer calibration ===\n");

		t0 = get_time();
		t1 = get_time();
		ccprintf("- back-to-back get_time : %d us\n", difftime(t0, t1));

		/* Sleep for 5 seconds */
		ccprintf("- sleep 1s :\n  ");
		cflush();
		ccprintf("Go...");
		t0 = get_time();
		crec_usleep(1000000);
		t1 = get_time();
		ccprintf("done. delay = %d us\n", difftime(t0, t1));

		/* try small usleep */
		ccprintf("- short sleep :\n");
		cflush();
		for (d = 128; d > 0; d = d / 2) {
			t0 = get_time();
			crec_usleep(d);
			t1 = get_time();
			ccprintf("  %d us => %d us\n", d, difftime(t0, t1));
			cflush();
		}

		ccprintf("Done.\n");
	}

	return EC_SUCCESS;
}

void run_test(int argc, const char **argv)
{
	task_wake(TASK_ID_TESTTMR);
}

File: test/timer_calib.py
# Copyright 2011 The Chromium OS Authors. All rights reserved.
# Use of this source code is governed by a BSD-style license that can be
# found in the LICENSE file.
#
# Check timers behavior
#

import time

def one_pass(helper):
      helper.wait_output("=== Timer calibration ===")
      res = helper.wait_output("back-to-back get_time : (?P<lat>[0-9]+) us",
                               use_re=True)["lat"]
      minlat = int(res)
      helper.trace("get_time latency %d us\n" % minlat)

      helper.wait_output("sleep 1s")
      t0 = time.time()
      second = helper.wait_output("done. delay = (?P<second>[0-9]+) us",
                               use_re=True)["second"]
      t1 = time.time()
      secondreal = t1 - t0
      secondlat = int(second) - 1000000
      helper.trace("1s timer latency %d us / real time %f s\n" % (secondlat,
                                                                  secondreal))


      us = {}
      for pow2 in range(7):
          delay = 1 << (7-pow2)
          us[delay] = helper.wait_output("%d us => (?P<us>[0-9]+) us" % delay,
                                  use_re=True)["us"]
      helper.wait_output("Done.")

      return minlat, secondlat, secondreal


def test(helper):
      one_pass(helper)

      helper.ec_command("reboot")
      helper.wait_output("--- UART initialized")

      # get the timing results on the second pass
      # to avoid binary translation overhead
      minlat, secondlat, secondreal = one_pass(helper)

      # check that the timings somewhat make sense
      if minlat > 220 or secondlat > 500 or abs(secondreal-1.0) > 0.200:
           helper.fail("imprecise timings " +
                       "(get_time %d us sleep %d us / real time %.3f s)" %
                       (minlat, secondlat, secondreal))

      return True # PASS !
